core/src/worker/activities.rs

```diff
@@ -514,6 +572,54 @@ pub(crate) struct TrackedPermittedTqResp {
     pub resp: PollActivityTaskQueueResponse,
 }
 
+struct GracefulShutdown {
+    once: TakeCell<GSState>,
+}
+struct GSState {
+    start_tasks_stream_complete: CancellationToken,
+    cancels_tx: UnboundedSender<PendingActivityCancel>,
+    grace_period: Duration,
+    outstanding: OutstandingActMap,
+}
+impl GracefulShutdown {
+    fn new(
+        grace_period: Duration,
+        start_tasks_stream_complete: CancellationToken,
+        cancels_tx: UnboundedSender<PendingActivityCancel>,
+        outstanding: OutstandingActMap,
+    ) -> Self {
+        Self {
+            once: TakeCell::new(GSState {
+                grace_period,
+                start_tasks_stream_complete,
+                cancels_tx,
+                outstanding,
+            }),
+        }
+    }
+
+    fn initiate(&self) {
+        if let Some(state) = self.once.take_once() {
+            let grace_deadline = Instant::now() + state.grace_period;
+            // We don't really need to join this
+            tokio::task::spawn(async move {
```

Comment:
I would make this a stream like the others and not have to worry about spawning this in the background or executing exactly once.

Reply:
Ah, sure good point